File: windows/src/OmniWheelPC/Protocol.cs
namespace OmniWheelPC.Network;

/// <summary>
/// Optimized binary protocol v2.
/// Supports both v1 (16B header) and v2 (8B header) for backward compatibility.
/// CRC-16/CCITT-FALSE with pre-computed lookup table.
/// </summary>
public static class Protocol
{
    public const byte ProtocolVersion = 2;
    public const ushort Magic = 0x4F57;

    public const int DiscoveryPort = 19700;
    public const int InputPort = 19701;
    public const int ControlPort = 19702;

    // v2 header size
    public const int HeaderSizeV2 = 8;
    // v1 header size (legacy)
    public const int HeaderSizeV1 = 16;
    public const int CrcSize = 2;

    public enum PacketType : byte
    {
        Discover = 0x01,
        DiscoverResponse = 0x02,
        Connect = 0x03,
        ConnectAck = 0x04,
        Disconnect = 0x05,
        Input = 0x06,
        Heartbeat = 0x07,
        HeartbeatAck = 0x08,
        Ping = 0x09,
        Pong = 0x0A,
        Error = 0xFF
    }

    // Input payload offsets
    public const int OffSteering = 0;
    public const int OffThrottle = 2;
    public const int OffBrake = 3;
    public const int OffClutch = 4;
    public const int OffGyroX = 5;
    public const int OffGyroY = 7;
    public const int OffGyroZ = 9;
    public const int ButtonBytes = 2; // always 2 bytes for up to 16 buttons

    // ========== CRC-16/CCITT-FALSE LOOKUP TABLE ==========
    private static readonly ushort[] CrcTable = new ushort[256];

    static Protocol()
    {
        for (int i = 0; i < 256; i++)
        {
            ushort crc = (ushort)(i << 8);
            for (int j = 0; j < 8; j++)
            {
                if ((crc & 0x8000) != 0)
                    crc = (ushort)((crc << 1) ^ 0x1021);
                else
                    crc = (ushort)(crc << 1);
            }
            CrcTable[i] = crc;
        }
    }

    public static ushort ComputeCrc(byte[] data, int offset, int length)
    {
        ushort crc = 0xFFFF;
        for (int i = offset; i < offset + length; i++)
        {
            crc = (ushort)((crc << 8) ^ CrcTable[((crc >> 8) ^ data[i]) & 0xFF]);
        }
        return crc;
    }

    // ========== PACKET BUILDING (v2 compact) ==========

    public static byte[] BuildPacket(PacketType type, ushort sequence, byte[]? payload = null)
    {
        int pLen = payload?.Length ?? 0;
        int totalLen = HeaderSizeV2 + pLen + CrcSize;
        var pkt = new byte[totalLen];

        // Magic
        pkt[0] = (byte)(Magic & 0xFF);
        pkt[1] = (byte)((Magic >> 8) & 0xFF);
        // Version
        pkt[2] = ProtocolVersion;
        // Type
        pkt[3] = (byte)type;
        // Sequence (1 byte)
        pkt[4] = (byte)(sequence & 0xFF);
        // Timestamp: ms within current second
        int ms = (int)(DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() % 60000);
        pkt[5] = (byte)(ms & 0xFF);
        pkt[6] = (byte)((ms >> 8) & 0xFF);
        // Payload length
        pkt[7] = (byte)(pLen & 0xFF);

        if (payload != null && pLen > 0)
            Array.Copy(payload, 0, pkt, HeaderSizeV2, pLen);

        // CRC
        var crc = ComputeCrc(pkt, 0, HeaderSizeV2 + pLen);
        pkt[HeaderSizeV2 + pLen] = (byte)(crc & 0xFF);
        pkt[HeaderSizeV2 + pLen + 1] = (byte)((crc >> 8) & 0xFF);

        return pkt;
    }

    // ========== PARSING (supports v1 and v2) ==========

    public struct ParsedHeader
    {
        public PacketType Type;
        public byte Sequence;
        public int TimestampMs;
        public int PayloadLength;
        public int HeaderSize; // 8 for v2, 16 for v1
    }

    /// <summary>
    /// Parse header supporting both v1 (16B) and v2 (8B) formats.
    /// Detects version from byte [2].
    /// </summary>
    public static bool TryParseHeader(byte[] data, int offset, out ParsedHeader header)
    {
        header = default;
        if (data == null || data.Length - offset < HeaderSizeV2) return false;

        ushort magic = (ushort)(data[offset] | (data[offset + 1] << 8));
        if (magic != Magic) return false;

        byte version = data[offset + 2];
        int headerSize;

        if (version == 2)
        {
            // v2: compact 8-byte header
            headerSize = HeaderSizeV2;
            header.Type = (PacketType)data[offset + 3];
            header.Sequence = data[offset + 4];
            header.TimestampMs = data[offset + 5] | (data[offset + 6] << 8);
            header.PayloadLength = data[offset + 7];
        }
        else if (version == 1)
        {
            // v1: legacy 16-byte header
            if (data.Length - offset < HeaderSizeV1) return false;
            headerSize = HeaderSizeV1;
            header.Type = (PacketType)data[offset + 3];
            header.Sequence = (byte)(data[offset + 4] | (data[offset + 5] << 8));
            header.TimestampMs = (int)(data[offset + 6] | (data[offset + 7] << 8) | (data[offset + 8] << 16) | (data[offset + 9] << 24));
            header.PayloadLength = data[offset + 10] | (data[offset + 11] << 8) | (data[offset + 12] << 16) | (data[offset + 13] << 24);
        }
        else
        {
            return false;
        }

        header.HeaderSize = headerSize;
        return true;
    }

    public static bool ValidateCrc(byte[] packet, int headerSize)
    {
        if (packet.Length < headerSize + CrcSize) return false;
        var crc = ComputeCrc(packet, 0, packet.Length - CrcSize);
        var received = (ushort)(packet[packet.Length - 2] | (packet[packet.Length - 1] << 8));
        return crc == received;
    }
}

File: android/app/src/main/java/com/topeck/omniwheel/network/Protocol.kt
package com.topeck.omniwheel.network

import java.net.DatagramPacket
import java.nio.ByteBuffer
import java.nio.ByteOrder

/**
 * Optimized binary protocol v2.
 *
 * **Compact header** (8 bytes, was 16):
 *   [0-1]  Magic       0x4F57 ("OW")
 *   [2]    Version     2
 *   [3]    Type        packet type
 *   [4]    Sequence    1-byte wrapping counter
 *   [5-6]  Timestamp   ms within current second (0..59999)
 *   [7]    PayloadLen  0..255
 *
 * **Input payload** (5 or 11 bytes, was always 11+):
 *   [0-1]  Steering    int16
 *   [2]    Throttle    uint8
 *   [3]    Brake       uint8
 *   [4]    Clutch      uint8
 *   [5-10] Gyro X/Y/Z  3x int16 (ONLY when gyro flag bit is set)
 *   [11+]  Buttons     bitmap (always 2 bytes for up to 16 buttons)
 *
 * **Packet**: [Header 8B] [Payload NB] [CRC-16 2B]
 *
 * Bandwidth at 240Hz:
 *   No gyro:  (8 + 5 + 2 + 2) * 240 = 4,128 B/s  (was ~7,440)
 *   With gyro: (8 + 11 + 2 + 2) * 240 = 6,912 B/s  (was ~7,440)
 *   With delta (no change): 0 B/s (up to ~80% reduction)
 */
object Protocol {
    const val PROTOCOL_VERSION: Byte = 2
    const val MAGIC: Short = 0x4F57

    // Ports
    const val DISCOVERY_PORT = 19700
    const val INPUT_PORT = 19701
    const val CONTROL_PORT = 19702

    const val HEADER_SIZE = 8
    const val CRC_SIZE = 2

    // Packet types
    const val TYPE_DISCOVER = 0x01.toByte()
    const val TYPE_DISCOVER_RESPONSE = 0x02.toByte()
    const val TYPE_CONNECT = 0x03.toByte()
    const val TYPE_CONNECT_ACK = 0x04.toByte()
    const val TYPE_INPUT = 0x06.toByte()
    const val TYPE_HEARTBEAT = 0x07.toByte()
    const val TYPE_HEARTBEAT_ACK = 0x08.toByte()
    const val TYPE_DISCONNECT = 0x05.toByte()
    const val TYPE_META = 0x0B.toByte()
    const val TYPE_HUD_WIDGET = 0x0C.toByte()
    const val TYPE_HUD_FULL = 0x0D.toByte()

    // Input payload offsets
    const val OFF_STEERING = 0
    const val OFF_THROTTLE = 2
    const val OFF_BRAKE = 3
    const val OFF_CLUTCH = 4
    const val OFF_GYRO_X = 5
    const val OFF_GYRO_Y = 7
    const val OFF_GYRO_Z = 9
    const val OFF_BUTTONS_NO_GYRO = 5   // buttons start here when no gyro
    const val OFF_BUTTONS_WITH_GYRO = 11 // buttons start here when gyro
    const val BUTTON_BYTES = 3            // always 3 bytes (up to 24 buttons)

    @Volatile
    var sequenceNumber: Short = 0
        private set

    @Synchronized
    fun nextSequence(): Byte {
        sequenceNumber = if (sequenceNumber >= 127) 0 else (sequenceNumber + 1).toShort()
        return sequenceNumber.toByte()
    }

    // ========== CRC-16/CCITT-FALSE LOOKUP TABLE ==========
    private val CRC_TABLE = ShortArray(256)

    init {
        for (i in 0..255) {
            var crc = i shl 8
            for (j in 0..7) {
                crc = if ((crc and 0x8000) != 0) (crc shl 1) xor 0x1021 else crc shl 1
            }
            CRC_TABLE[i] = (crc and 0xFFFF).toShort()
        }
    }

    fun computeCrc16(data: ByteArray, offset: Int, length: Int): Short {
        var crc = 0xFFFF
        for (i in offset until offset + length) {
            val idx = ((crc ushr 8) xor (data[i].toInt() and 0xFF)) and 0xFF
            crc = ((crc shl 8) xor CRC_TABLE[idx].toInt()) and 0xFFFF
        }
        return crc.toShort()
    }

    // ========== PACKET BUILDING ==========

    fun buildPacket(type: Byte, payload: ByteArray? = null): ByteArray {
        val pLen = payload?.size ?: 0
        val seq = nextSequence()
        val totalLen = HEADER_SIZE + pLen + CRC_SIZE

        // Use thread-local buffer pool: allocate once, reuse
        val buf = ByteArray(totalLen)

        // Header (8 bytes)
 buf[0] = (MAGIC.toInt() and 0xFF).toByte()
        buf[1] = ((MAGIC.toInt() shr 8) and 0xFF).toByte()
        buf[2] = PROTOCOL_VERSION
        buf[3] = type
        buf[4] = seq
        // Timestamp: ms within current second
        val msInSecond = ((System.currentTimeMillis() % 60000).toInt() and 0xFFFF)
        buf[5] = (msInSecond and 0xFF).toByte()
        buf[6] = ((msInSecond shr 8) and 0xFF).toByte()
        buf[7] = (pLen and 0xFF).toByte()

        // Payload
        if (payload != null && pLen > 0) {
            System.arraycopy(payload, 0, buf, HEADER_SIZE, pLen)
        }

        // CRC
        val crcVal = computeCrc16(buf, 0, HEADER_SIZE + pLen)
        buf[HEADER_SIZE + pLen] = (crcVal.toInt() and 0xFF).toByte()
        buf[HEADER_SIZE + pLen + 1] = ((crcVal.toInt() shr 8) and 0xFF).toByte()

        return buf
    }

    /**
     * Build optimized input packet.
     * @param includeGyro if true, includes 6 bytes of gyro data
     * @param steering raw steering value
     * @param throttle 0..255
     * @param brake 0..255
     * @param clutch 0..255
     * @param gyroX/Y/Z gyro axes
     * @param activeButtons button IDs that are pressed
     */
    fun buildInputPacket(
        steering: Short, throttle: Byte, brake: Byte, clutch: Byte,
        gyroX: Short = 0, gyroY: Short = 0, gyroZ: Short = 0,
        activeButtons: Set<Int> = emptySet(),
        includeGyro: Boolean = false
    ): ByteArray {
        val gyroBytes = if (includeGyro) 6 else 0
        val payloadSize = 5 + gyroBytes + BUTTON_BYTES
        val payload = ByteArray(payloadSize)

        // Steering
        payload[0] = (steering.toInt() and 0xFF).toByte()
        payload[1] = ((steering.toInt() shr 8) and 0xFF).toByte()
        // Throttle, Brake, Clutch
        payload[2] = throttle
        payload[3] = brake
        payload[4] = clutch

        // Gyro (conditional)
        if (includeGyro) {
            payload[5] = (gyroX.toInt() and 0xFF).toByte()
            payload[6] = ((gyroX.toInt() shr 8) and 0xFF).toByte()
            payload[7] = (gyroY.toInt() and 0xFF).toByte()
            payload[8] = ((gyroY.toInt() shr 8) and 0xFF).toByte()
            payload[9] = (gyroZ.toInt() and 0xFF).toByte()
            payload[10] = ((gyroZ.toInt() shr 8) and 0xFF).toByte()
        }

        // Button bitmap (always 3 bytes)
        val btnOffset = 5 + gyroBytes
        var bitmap = 0
        for (btn in activeButtons) {
            val bitIdx = (btn - 1) // 0-indexed bit
            if (bitIdx in 0..23) bitmap = bitmap or (1 shl bitIdx)
        }
        payload[btnOffset] = (bitmap and 0xFF).toByte()
        payload[btnOffset + 1] = ((bitmap shr 8) and 0xFF).toByte()

        return buildPacket(TYPE_INPUT, payload)
    }

    fun buildDiscoverPacket(deviceName: String): ByteArray {
        return buildPacket(TYPE_DISCOVER, deviceName.toByteArray(Charsets.UTF_8))
    }

    fun buildHeartbeatPacket(): ByteArray {
        return buildPacket(TYPE_HEARTBEAT)
    }

    /**
     * Build a metadata packet describing the phone:
     *   [0] battery % (0..100, 255 = unknown)
     *   [1-2] steering max angle uint16 LE
     *   [3-6] screen width px uint32 LE
     *   [7-10] screen height px uint32 LE
     *   [11] flags (bit0 = clutch enabled)
     *   [12+] device type name UTF-8
     */
    fun buildMetaPacket(
        batteryPercent: Int,
        maxAngle: Int,
        screenWidthPx: Int,
        screenHeightPx: Int,
        deviceType: String,
        clutchEnabled: Boolean = true
    ): ByteArray {
        val nameBytes = deviceType.toByteArray(Charsets.UTF_8)
        val payload = ByteArray(12 + nameBytes.size)
        payload[0] = batteryPercent.coerceIn(0, 100).toByte()
        // max angle u16 LE
        val ma = maxAngle.coerceIn(0, 65535)
        payload[1] = (ma and 0xFF).toByte()
        payload[2] = ((ma shr 8) and 0xFF).toByte()
        // screen w u32 LE
        payload[3] = (screenWidthPx and 0xFF).toByte()
        payload[4] = ((screenWidthPx shr 8) and 0xFF).toByte()
        payload[5] = ((screenWidthPx shr 16) and 0xFF).toByte()
        payload[6] = ((screenWidthPx shr 24) and 0xFF).toByte()
        // screen h u32 LE
        payload[7] = (screenHeightPx and 0xFF).toByte()
        payload[8] = ((screenHeightPx shr 8) and 0xFF).toByte()
        payload[9] = ((screenHeightPx shr 16) and 0xFF).toByte()
        payload[10] = ((screenHeightPx shr 24) and 0xFF).toByte()
        // flags
        payload[11] = if (clutchEnabled) 1 else 0
        nameBytes.copyInto(payload, 12)
        return buildPacket(TYPE_META, payload)
    }

    // ========== PARSING ==========

    data class ParsedHeader(
        val type: Byte,
        val sequence: Byte,
        val timestampMs: Int,
        val payloadLength: Int
    )

    fun parseHeader(data: ByteArray, offset: Int = 0): ParsedHeader? {
        if (data.size - offset < HEADER_SIZE) return null
        val magic = ((data[offset + 1].toInt() and 0xFF) shl 8) or (data[offset].toInt() and 0xFF)
        if (magic != MAGIC.toInt()) return null
        val version = data[offset + 2]
        // Accept v1 and v2 for backward compatibility during transition
        if (version != 1.toByte() && version != PROTOCOL_VERSION) return null
        return ParsedHeader(
            type = data[offset + 3],
            sequence = data[offset + 4],
            timestampMs = (data[offset + 5].toInt() and 0xFF) or
                         ((data[offset + 6].toInt() and 0xFF) shl 8),
            payloadLength = data[offset + 7].toInt() and 0xFF
        )
    }

    fun validateCrc(packet: ByteArray): Boolean {
        if (packet.size < HEADER_SIZE + CRC_SIZE) return false
        val crcDataLen = packet.size - CRC_SIZE
        val computed = computeCrc16(packet, 0, crcDataLen)
        val received = ((packet[packet.size - 1].toInt() and 0xFF) shl 8) or
                       (packet[packet.size - 2].toInt() and 0xFF)
        return (computed.toInt() and 0xFFFF) == received
    }
}
